northstar/src/runtime/island/init.rs

```diff
@@ -0,0 +1,586 @@
+                        exit(code);
+                    }
+                    // TODO: Other waitpid results
+                    _ => panic!("abnormal exit of child process"),
+                };
+            }
+            unistd::ForkResult::Child => {
+                // Wait until init is prepared for us to crash (4)
+                cond_init.wait();
+
+                drop(intercom);
+                reset_signal_handlers();
+                set_parent_death_signal(SIGKILL).expect("Failed to set parent death signal");
+
+                let (init, argv, env) = args(&container.manifest);
+                println!("{} init: {:?}", id, init);
+                println!("{} argv: {:#?}", id, argv);
+                println!("{} env: {:#?}", id, env);
+
+                // The pipe in cond_child is closed on execve (5)
+                panic!("{}: {:?}", id, unistd::execve(&init, &argv, &env))
+            }
+        },
+        Err(e) => panic!("Fork error: {}", e),
+    }
+}
+
+/// Prepare the environment in init
+fn prepare(
+    config: &Config,
+    container: &Container<IslandProcess>,
+    fds: HashMap<i32, i32>,
+    supplementary_groups: &[(String, Option<Gid>)],
+) -> anyhow::Result<()> {
+    let manifest = &container.manifest;
+    let root = container.root.canonicalize()?;
+
+    // Install signal handlers that forward every signal to our child
+    init_signal_handlers();
+
+    // Mount
+    rootfs(&config, &container).context("Failed to mount")?;
+
+    // Chroot
+    cdebug!("Setting chroot to {}", root.display());
+    unistd::chroot(&root).context("Failed to chroot")?;
+
+    // Pwd
+    cdebug!("Setting pwd to /");
+    env::set_current_dir("/").context("Failed to set cwd to /")?;
+
+    // UID / GID
+    setid(manifest.uid, manifest.gid).context("Failed to setuid/gid")?;
+
+    // Supplementary groups
+    setgroups(manifest.gid, &supplementary_groups)?;
+
+    cdebug!("Setting no new privs");
+    set_no_new_privs(true)?;
+
+    // Become a subreaper for orphans in this namespace
+    cdebug!("Setting child subreaper flag");
+    set_child_subreaper(true)?;
+
+    // Set the parent process death signal of the calling process to arg2
+    // (either a signal value in the range 1..maxsig, or 0 to clear).
+    cdebug!("Setting parent death signal to SIGKILL");
+    set_parent_death_signal(SIGKILL)?;
+
+    // Capabilities
+    drop_capabilities(manifest.capabilities.as_ref()).context("Failed to drop privs")?;
+
+    close_file_descriptors(fds)?;
+
+    // We cannot use log after here because the fd to logd is closed on Android
+
+    Ok(())
+}
+
+// TODO: The container could be malformed and the mountpoint might be
+// missing. This is not a fault of the RT so don't expect it.
+// TODO: mount flags: nosuid etc....
+// TODO: /dev mounts from manifest: full or minimal
+fn rootfs(config: &Config, container: &Container<IslandProcess>) -> anyhow::Result<()> {
```

Comment:
the whole `init` would benifit from `thiserror` Error definitions

Reply:
This would be just boilerplate since the errors are never passed outside of `init`. Using anyhow is convenient because the only thing what could be done with those is printing the in `init.